### ISSUE TITLE
Focus(TVOS)

### ISSUE THREAD

**Issue by sebasg0** [react-native-tvos/react-native-tvos#388]
### Description

In a catalog of products (multiple horizontal flatlists inside of a vertical flatlist), I will get delayed focus behavior in TVOS while switching in between elements, focus animation comes delayed. This got me thinking if somehow we could create a native component with the focus logic and manage the toggle style logic in within the native component which is running in main thread rather than using setState to manage it as TouchableOpacity does. Has someone tried this? is so it will be really helpful to know.

### Version

0.69

### Output of `npx react-native info`

System:
    OS: macOS 12.6
    CPU: (10) arm64 Apple M1 Pro
    Memory: 445.08 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.10.0 - /opt/homebrew/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 6.14.16 - /usr/local/bin/npm
    Watchman: 2022.09.26.00 - /opt/homebrew/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.1, iOS 16.1, macOS 13.0, tvOS 16.1, watchOS 9.1
    Android SDK: Not Found
  IDEs:
    Android Studio: 2021.3 AI-213.7172.25.2113.9014738
    Xcode: 14.1/14B47b - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.14.1 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.0.0 => 18.0.0 
    react-native: Not Found
    react-native-macos: Not Found
    react-native-tvos:  0.69.6-0 
  npmGlobalPackages:
    *react-native*: Not Found

### Steps to reproduce

Create any project with a catalog of products and render at least 400 elements.

**Comment by sebasg0** [comment]
@douglowder do you have any suggestions for this behavior?

**Comment by wouterds** [comment]
This is indeed a pain point, the focus and reflecting it in the UI through state. Especially as apps grow and get bigger, fetch more data (even in background) it becomes very very very hard to maintain a decent user experience.

**Comment by douglowder** [comment]
Possibly we need to look at using the native focus APIs that are provided for collection views. Also hoping that Fabric will solve some of these issues.

If one of you has an example that shows this problem, I can do some performance testing to see where the bottlenecks are that are slowing things down.

**Comment by sebasg0** [comment]
Issue is with large data sets, rendering a 20 x 20 nested flatlist/flashlist and then adding some logic to every card for focus style, this way you will be able to reproduce the delayed focus. I can see the parallax effect is already in the card but since the style is happening in the JS thread then it's coming delayed. This is why I was asking if we could somehow create a native component and inject a React view as children and handle the style toggle on focus/blur in the native thread.


@douglowder I'm not very experienced in IOS but I know this could be achieved in android. Do you have any code snipet we could try of a UIView with focus logic natively in swift?

**Comment by douglowder** [comment]
Interesting... if you disable the parallax animations, does the problem resolve?

**Comment by sebasg0** [comment]
https://user-images.githubusercontent.com/83782787/200992800-6d073ff4-f8a4-4c61-b72b-a58fe930bc0e.mov

**Comment by sebasg0** [comment]
Here is visible how the parallax effect is already in the last card but the focus comes delayed, I'm wrapping my set state inside a `requestAnimationFrame` as well

**Comment by sebasg0** [comment]
Update: I was able to create a native wrapper, just exploring some props and how far we could take it. Once I got a stable version I will post it with you guys. This will solve our focus issues when having a large set of elements in a flatlist in tvOS/androidtv/firestick.

https://user-images.githubusercontent.com/83782787/201027806-ae0fcaa2-f032-4ddd-9f23-c0b362d465a5.mov

**Comment by wouterds** [comment]
An optional focusStyle prop would also be a solution. That way we don't need to manage state to adjust styles when a component gets focus.

**Comment by sebasg0** [comment]
@wouterds I believe this will only be possible by passing an style object but limited one with props as opacity, borderRadius, paddings, margins, we could add also an animation prop as appletv streaming apps have this scale anim, etc. What properties will you suggest? for me right now, passing the animation, border radius and border color it's enough.

**Comment by douglowder** [comment]
@sebasg0 @wouterds when a focusStyle object is used, should parallax animations be turned off by default?

**Comment by sebasg0** [comment]
No, we can leave it enabled with no issues. We just need to sort of the nextFocus prop in this new wrapper, blocking focus is way easier natively rather than passing a reference.

**Comment by wouterds** [comment]
> @wouterds I believe this will only be possible by passing an style object but limited one with props as opacity, borderRadius, paddings, margins, we could add also an animation prop as appletv streaming apps have this scale anim, etc. What properties will you suggest? for me right now, passing the animation, border radius and border color it's enough.

@sebasg0 Why not allow all styles? I would see it as a secondary style object that just enriches the `style` object if it is provided, on focus.

**Comment by sebasg0** [comment]
@wouterds I was stuck in some mobile tasks at my company. I will re take this today. First stable version is ready with a catalog example project which is running very performant in appleTV HD. I will try to publish package to npm today so u'all can give feedback and contribute accordingly. We are still missing support for AndroidTV. Also I made a concerning discovery in react drawer navigation... It has many focus issues, only way is to go with a native-stack, as many apps need a side bar menu we will have to come up with a solution for this as well.
